### PR TITLE
[CommonOperatorLoader, Interpreter and CPU Backend] Add support for ReduceProd Operator

### PR DESCRIPTION
**Pull request by vbaddi**
Summary:
Adding support for ONNX ReduceProd operator.
Also, Added Interpreter and CPU Backend Implementation.

Documentation: None

Test Plan: Ninja test is run

### CODE REVIEW

**Comment by vbaddi** [comment]
@jfix71, Could you please review this changes?
Sorry for the late response, earlier i had raise a PR [#4863](https://github.com/pytorch/glow/pull/4863) for the same, but it is closed now.
One of the review comments was to merge the implementation of ReduceAdd and ReduceProd, since most of the code in InterpreterNodes.cpp is common. I have made those changes. Let me know your thoughts.

**Comment by vuzelac-cadence** [comment]
Hi @vbaddi , I have this implemented but didn't upstream.
I suggest not doing the macro but templatized functors to overload the operator.

Also, you are sharing the quantized implementation although I don't see  the test for it ? Quantization for Prod is much more complex than for Add.

**Comment by vbaddi** [comment]
@vuzelac-cadence 
Thank you for the suggestions.
I see that ReduceMin and ReduceMax operators also handled with MACROS, so i followed similar to them,
Regarding the Quantization for Prod, i agree it is bit complex then add, But Currently i tried to keep the format same as ReduceAdd and we can address the same on need basis.

**Comment by vbaddi** [comment]
@jfix71, Could you please review this changes?

**Comment by vuzelac-cadence** [comment]
Hi @vbaddi - for the quantization, what error, message, warning, user will get ? The way I see it, it all will run fun just with invalid quantized results, thus, should not have it supported.

**Comment by vbaddi** [comment]
@vuzelac-cadence, Yes i agree,
Will try to add error codes for the same.
Or currently will support FP32 precision only.

**Comment by vuzelac-cadence** [comment]
I suggest removing the quantized part (the if part, and leave the else) and just have there this:
assert(!batch->getType()->isQuantizedType() &&
           "Quantized implementation not supported yet.");

**Comment by vbaddi** [comment]
@vuzelac-cadence,
Thank you.
Have updated the patch. Let me know if any more changes.

**Review by jfix71 — CHANGES_REQUESTED** [review]
@vbaddi I'm a bit confused here -- you're blacklisting tests e.g. for float16 that are not actually present in OperatorTest. Can you clarify what exactly is supposed to be supported and update the diff accordingly -- both in the blacklists in each `[Backend]OperatorTest.cpp`, and in `Backend::isOpSupported()`?

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 1]
```diff
@@ -3712,39 +3712,47 @@ void BoundInterpreterFunction::fwdBatchedAddInst(
...
-          }
-        }
-      }
-    }
-  }
-}
+/// Define fwdBatchedReduceAddInstImpl
+DEFINE_REDUCEADDPROD_INST_IMPL(ReduceProd, 1, *)
 
+#undef DEFINE_REDUCEADDPROD_INST_IMPL
```
[the remark this answers is not in the document]
Comment (vbaddi) [reply lib/Backends/Interpreter/InterpreterNodes.cpp:4074]: Done

File: lib/Backends/NNPI/tests/NNPIOperatorTest.cpp [hunk 2]
```diff
@@ -490,4 +496,4 @@ struct BlacklistInitializer {
     TestBlacklist::prepareBlacklist(testBlacklistedSetups,
                                     backendTestBlacklist);
   }
-} blacklistInitializer;
+} blacklistInitializer;
```
[the remark this answers is not in the document]
Comment (vbaddi) [reply lib/Backends/NNPI/tests/NNPIOperatorTest.cpp]: Done

**Comment by vbaddi** [comment]
@jfix71, I have modified the support list changes accordingly.
Could you please review the changes?

File: lib/Backends/Interpreter/Interpreter.cpp [hunk 3]
```diff
@@ -80,6 +80,7 @@ bool Interpreter::isOpSupported(const NodeInfo &NI) const {
   switch (NI.getKind()) {
   case Kinded::Kind::BatchedReduceMinNodeKind:
   case Kinded::Kind::BatchedReduceMaxNodeKind:
+  case Kinded::Kind::BatchedReduceProdNodeKind:
```
Comment (jfix71) [inline lib/Backends/Interpreter/Interpreter.cpp:83]: Here you're claiming support for Float16 and BFloat16, but there is no test coverage in OperatorTest. Can you add a couple extra tests to cover this?

File: lib/Backends/CPU/tests/CPUOperatorTest.cpp [hunk 4]
```diff
@@ -46,6 +46,11 @@ std::set<std::string> glow::backendTestBlacklist = {
     "batchedPairwiseDotProduct/0",
     "batchedReduceAdd_BFloat16/0",
     "batchedReduceAdd_Float16/0",
+    "batchedReduceProd_Float16/0",
+    "batchedReduceProd_BFloat16/0",
+    "batchedReduceProd_Int32/0",
+    "batchedReduceProd_Int64/0",
+    "batchedReduceProd_Int8/0",
```
Comment (jfix71) [inline lib/Backends/CPU/tests/CPUOperatorTest.cpp:53]: Still confused here -- there aren't any such `Float16`/`BFloat16`/`Int8` tests in OperatorTests. Can you make sure that we have sufficient coverage in OperatorTests to cover new functionality/kernels in all backends, and then only blacklist the actually existing tests?

File: lib/Backends/Interpreter/Interpreter.cpp (hunk 3 above)
Comment (vbaddi) [reply lib/Backends/Interpreter/Interpreter.cpp:83]: Added test cases for Float16 and BFloat16 in OperatorTest.cpp

File: lib/Backends/CPU/tests/CPUOperatorTest.cpp (hunk 4 above)
Comment (vbaddi) [reply lib/Backends/CPU/tests/CPUOperatorTest.cpp:53]: CPU supports only Float32, Added the changes accordingly.

**Comment by vbaddi** [comment]
@jfix71, Could you please review this changes and merge?

**Review by jfix71 — APPROVED** [review]
🚢

**Comment by jfix71** [comment]
@vbaddi I'm actually hitting some merge conflicts here due to `InterpreterNodes.cpp` in https://github.com/pytorch/glow/pull/5081, CC: @sgeorgiev255. Would you mind resolving this? Please tag me in your response so I am sure to see it, thanks.

**Comment by vbaddi** [comment]
> @vbaddi I'm actually hitting some merge conflicts here due to `InterpreterNodes.cpp` in #5081, CC: @sgeorgiev255. Would you mind resolving this? Please tag me in your response so I am sure to see it, thanks.

@jfix71, Could you please merge this changes? Conflicts are resolved.

**Comment by vbaddi** [comment]
@jfix71, Could you please merge this changes?

**Review by jfix71 — APPROVED** [review]
Sorry for the delay!

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 5]
```diff
@@ -6811,4 +6846,4 @@ void BoundInterpreterFunction::fwdBBoxTransformInst(
 void BoundInterpreterFunction::fwdExternalFunctionCallInst(
     glow::ExternalFunctionCallInst const *) {
   LOG(FATAL) << "ExternalFunctionCallInst is not supported yet";
-}
+}
```
Comment (jfix71) [inline lib/Backends/Interpreter/InterpreterNodes.cpp:6849]: @vbaddi Sorry one last thing -- can you add back the newline here? Our internal linters don't like this and it's causing me some issues.

Comment (vbaddi) [reply lib/Backends/Interpreter/InterpreterNodes.cpp]: Done

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 6]
```diff
@@ -1026,8 +1026,7 @@ void BoundInterpreterFunction::fwdBatchNormalizationFloatImpl(
   auto biasH = getWeightHandle<ElemTy>(I->getBias());
   auto meanH = getWeightHandle<ElemTy>(I->getMean());
   auto varH = getWeightHandle<ElemTy>(I->getVar());
-  unsigned_t channelIdx =
-      I->getChannelIdx(); // NOTE: We only support NTHWC, NHWC, NWC and NCW
+  unsigned_t channelIdx = I->getChannelIdx();
```
Comment (jfix71) [inline lib/Backends/Interpreter/InterpreterNodes.cpp:1029]: e.g. here

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 7]
```diff
@@ -1067,14 +1066,18 @@ void BoundInterpreterFunction::fwdBatchNormalizationFloatImpl(
...
     C = I->getSrc()->dims()[channelIdx];
     sizeImg = I->getSrc()->dims()[channelIdx == 2 ? 1 : 2];
     sizeN = C * sizeImg;
     isCMinor = (channelIdx == 2);
+  } else {
+    N = I->getSrc()->dims()[0];
+    C = I->getSrc()->dims()[channelIdx];
+    sizeImg = 1;
+    sizeN = C;
+    isCMinor = false;
```
Comment (jfix71) [inline lib/Backends/Interpreter/InterpreterNodes.cpp:1080]: e.g. here

**Review by jfix71 — COMMENTED** [review]
Hi @vbaddi -- sorry to come back to this again. It seems like there is some weird rebasing issue that occurred on your end. There are changes from previous commit(s) that appear to have been brought into this PR, but you aren't actually rebased on top of those commits as far as I can tell. So they appear as changes as part of this commit, and it's causing our internal tooling to break. I marked a couple places, related to batch normalization in InterpreterNodes.cpp (and there are lots of other changes just on comments/newlines as well).

Can you see what commit this PR is actually on top of, and make sure you're actually rebased on some pretty recent commit?

File: lib/Backends/Interpreter/InterpreterNodes.cpp (hunk 6 above)
Comment (vbaddi) [reply lib/Backends/Interpreter/InterpreterNodes.cpp]: @jfix71, Before rebasing, the last commit was #5195, now pulled in the latest commit #5300 and rebased on top of it.
Let me know if there are still any more conflicts.
